modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java

```diff
@@ -1028,4 +1087,357 @@ public Set<ReplicationGroupId> startedGroups() {
+                ReplicaLifecycleContext context = replicaContexts.get(parameters.groupId());
+
+                if (context != null) {
+                    synchronized (context) {
+                        context.reservedForPrimary = false;
+                    }
+                }
+            }
+
+            return falseCompletedFuture();
+        }
+
+        ReplicaLifecycleContext getContext(ReplicationGroupId groupId) {
+            return replicaContexts.computeIfAbsent(groupId,
+                    // Treat the absence in the map as STOPPED.
+                    k -> new ReplicaLifecycleContext(ReplicaState.STOPPED, nullCompletedFuture()));
+        }
+
+        /**
+         * Can possibly start replica if it's not running or is stopping.
+         *
+         * @param groupId Group id.
+         * @param startOperation Replica start operation.
+         * @param forcedAssignments Assignments to reset forcibly, if needed. Assignments reset is only available when replica is started.
+         */
+        CompletableFuture<Boolean> weakReplicaStart(
+                ReplicationGroupId groupId,
+                Supplier<CompletableFuture<Boolean>> startOperation,
+                @Nullable Assignments forcedAssignments
+        ) {
+            ReplicaLifecycleContext context = getContext(groupId);
+
+            synchronized (context) {
+                ReplicaState state = context.replicaState;
+
+                log.debug("Weak replica start [grp={}, state={}, future={}].", groupId, state, context.previousOperationFuture);
+
+                if (state == ReplicaState.STOPPED || state == ReplicaState.STOPPING) {
+                    return startReplica(groupId, context, startOperation);
+                } else if (state == ReplicaState.ASSIGNED) {
+                    if (forcedAssignments != null) {
+                        assert forcedAssignments.force() :
+                                format("Unexpected assignments to force [assignments={}, groupId={}].", forcedAssignments, groupId);
+
+                        replicaManager.resetPeers(groupId, fromAssignments(forcedAssignments.nodes()));
+                    }
+
+                    // Telling the caller that the replica is started.
+                    return trueCompletedFuture();
+                } else if (state == ReplicaState.PRIMARY_ONLY) {
+                    context.replicaState = ReplicaState.ASSIGNED;
+
+                    return trueCompletedFuture();
+                } // else no-op.
+
+                throw new AssertionError("Replica start cannot begin while the replica is being started [groupId=" + groupId + "].");
+            }
+        }
+
+        private CompletableFuture<Boolean> startReplica(
+                ReplicationGroupId groupId,
+                ReplicaLifecycleContext context,
+                Supplier<CompletableFuture<Boolean>> startOperation
+        ) {
+            context.replicaState = ReplicaState.STARTING;
```

Comment:
Why it's not implemented as a state machine? It might be simper to have ReplicaLifecycleStateMachine instead  ReplicaLifecycle with single public method changeState(groupId, newState), internally it will conditionally change the state and call the closure if needed. Some states should await another some will be applied instantly, some are considered as impossible and will through an exception.

Reply:
actually all the logic is in 2 methods now (weakStartReplica and weakStopReplica), maybe one single method would be too big; also it would have to accept the type of action (start or stop), because the transitions depend on it, so I am not sure the conditional logic will look more simple, just sheet of `if`-s in the number equal to cartesian product size